File: base-v3/src/main/java/com/adyen/checkout/base/model/payments/request/SepaPaymentMethod.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 22/8/2019.
 */

package com.adyen.checkout.base.model.payments.request;

import android.os.Parcel;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.adyen.checkout.base.util.PaymentMethodTypes;
import com.adyen.checkout.core.exception.ModelSerializationException;
import com.adyen.checkout.core.model.JsonUtils;

import org.json.JSONException;
import org.json.JSONObject;

@SuppressWarnings({"MemberName", "PMD.DataClass"})
public class SepaPaymentMethod extends PaymentMethodDetails {

    @NonNull
    public static final Creator<SepaPaymentMethod> CREATOR = new Creator<>(SepaPaymentMethod.class);

    public static final String PAYMENT_METHOD_TYPE = PaymentMethodTypes.SEPA;

    private static final String OWNER_NAME = "sepa.ownerName";
    private static final String IBAN_NUMBER = "sepa.ibanNumber";

    @NonNull
    public static final Serializer<SepaPaymentMethod> SERIALIZER = new Serializer<SepaPaymentMethod>() {

        @NonNull
        @Override
        public JSONObject serialize(@NonNull SepaPaymentMethod modelObject) {
            final JSONObject jsonObject = new JSONObject();
            try {
                // getting parameters from parent class
                jsonObject.putOpt(PaymentMethodDetails.TYPE, modelObject.getType());

                jsonObject.putOpt(OWNER_NAME, modelObject.getOwnerName());
                jsonObject.putOpt(IBAN_NUMBER, modelObject.getIbanNumber());
            } catch (JSONException e) {
                throw new ModelSerializationException(SepaPaymentMethod.class, e);
            }
            return jsonObject;
        }

        @NonNull
        @Override
        public SepaPaymentMethod deserialize(@NonNull JSONObject jsonObject) {
            final SepaPaymentMethod sepaPaymentMethod = new SepaPaymentMethod();

            // getting parameters from parent class
            sepaPaymentMethod.setType(jsonObject.optString(PaymentMethodDetails.TYPE, null));

            sepaPaymentMethod.setOwnerName(jsonObject.optString(OWNER_NAME, null));
            sepaPaymentMethod.setIbanNumber(jsonObject.optString(IBAN_NUMBER, null));

            return sepaPaymentMethod;
        }
    };

    private String ownerName;
    private String ibanNumber;

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        JsonUtils.writeToParcel(dest, SERIALIZER.serialize(this));
    }

    @Nullable
    public String getOwnerName() {
        return ownerName;
    }

    public void setOwnerName(@Nullable String ownerName) {
        this.ownerName = ownerName;
    }

    @Nullable
    public String getIbanNumber() {
        return ibanNumber;
    }

    public void setIbanNumber(@Nullable String ibanNumber) {
        this.ibanNumber = ibanNumber;
    }
}

File: components-core/src/main/java/com/adyen/checkout/components/core/paymentmethod/SepaPaymentMethod.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 22/8/2019.
 */
package com.adyen.checkout.components.core.paymentmethod

import com.adyen.checkout.components.core.PaymentMethodTypes
import com.adyen.checkout.core.exception.ModelSerializationException
import com.adyen.checkout.core.internal.data.model.getStringOrNull
import kotlinx.parcelize.Parcelize
import org.json.JSONException
import org.json.JSONObject

@Parcelize
class SepaPaymentMethod(
    override var type: String?,
    @Deprecated("This property is deprecated. Use the SERIALIZER to send the payment data to your backend.")
    override var checkoutAttemptId: String?,
    var ownerName: String?,
    var iban: String?,
) : PaymentMethodDetails() {

    companion object {
        const val PAYMENT_METHOD_TYPE = PaymentMethodTypes.SEPA
        private const val OWNER_NAME = "ownerName"
        private const val IBAN = "iban"

        @JvmField
        val SERIALIZER: Serializer<SepaPaymentMethod> = object : Serializer<SepaPaymentMethod> {
            override fun serialize(modelObject: SepaPaymentMethod): JSONObject {
                return try {
                    JSONObject().apply {
                        putOpt(TYPE, modelObject.type)
                        putOpt(CHECKOUT_ATTEMPT_ID, modelObject.checkoutAttemptId)
                        putOpt(OWNER_NAME, modelObject.ownerName)
                        putOpt(IBAN, modelObject.iban)
                    }
                } catch (e: JSONException) {
                    throw ModelSerializationException(SepaPaymentMethod::class.java, e)
                }
            }

            override fun deserialize(jsonObject: JSONObject): SepaPaymentMethod {
                return SepaPaymentMethod(
                    type = jsonObject.getStringOrNull(TYPE),
                    checkoutAttemptId = jsonObject.getStringOrNull(CHECKOUT_ATTEMPT_ID),
                    ownerName = jsonObject.getStringOrNull(OWNER_NAME),
                    iban = jsonObject.getStringOrNull(IBAN),
                )
            }
        }
    }
}
